basetest.pm

```diff
@@ -658,9 +665,13 @@ sub parse_serial_output_qemu ($self) {
                 elsif ($type =~ 'hard|fatal') {
                     $die = 1;
                     $fail_type = 'fail';
-                    $self->{fatal_failure} = $type eq 'fatal';
+                    $self->{fatal_failure} |= $type eq 'fatal';
                 }
                 $self->record_resultfile($message, $message . " - Serial error: $line", result => $fail_type);
+                # don't make overall result better - only worse
+                if (defined($self->{result})) {
+                    next if ($fail_severity->{$self->{result}} > $fail_severity->{$fail_type});
+                }
```

Comment:
You can avoid multiple ifs:
```suggestion
                next if defined($self->{result}) && $fail_severity->{$self->{result}} > $fail_severity->{$fail_type};
```

But you should also better avoid warnings and unintended updates in case the your mapping is insufficient:

```suggestion
                next if defined($self->{result}) && ($fail_severity->{$self->{result}} // 999) > $fail_severity->{$fail_type};
```

Reply:
The question is if the default should be 999 or 0 then

Reply:
I made the default 999 because it makes generally sense to preserve a "special" result outside the severity scale.